photon-client/src/main/scala/com/linkedin/photon/ml/data/DataValidators.scala

```diff
@@ -27,151 +26,129 @@ import com.linkedin.photon.ml.{DataValidationType, TaskType}
-  }
+  def nonNegativeLabels(labeledPoint: LabeledPoint): Boolean = labeledPoint.label >= 0
 
   /**
+   * Verify that the feature values of a data point are finite.
    *
-   * @param labeledPoint
-   * @return
+   * @param labeledPoint The input data point
+   * @return Whether all feature values for the input data point are finite
    */
-  def binaryLabel(labeledPoint: LabeledPoint): Boolean = {
-    BinaryClassifier.positiveClassLabel == labeledPoint.label ||
-      BinaryClassifier.negativeClassLabel == labeledPoint.label
-  }
+  def finiteFeatures(labeledPoint: LabeledPoint): Boolean =
+    labeledPoint.features.iterator.forall { case (_, value) =>
+      !(value.isNaN || value.isInfinite)
+    }
 
   /**
+   * Verify that a data point has a finite offset.
    *
-   * @param labeledPoint
-   * @return
+   * @param labeledPoint The input data point
+   * @return Whether the offset of the input data point is finite
    */
-  def finiteFeatures(labeledPoint: LabeledPoint): Boolean = {
-    var result = true
-    var idx = 0
-    while (idx < labeledPoint.features.activeSize && result) {
-      val value = labeledPoint.features match {
-        case d:DenseVector[Double] => d.valueAt(idx)
-        case s:SparseVector[Double] => s.valueAt(idx)
-      }
-      result = result && !(value.isNaN || value.isInfinite)
-      idx += 1
-    }
+  def finiteOffset(labeledPoint: LabeledPoint): Boolean = !(labeledPoint.offset.isNaN || labeledPoint.offset.isInfinite)
 
-    result
-  }
+  /**
+   * Verify that a data point has a finite weight.
+   *
+   * @param labeledPoint The input data point
+   * @return Whether the weight of the input data point is finite
+   */
+  def finiteWeight(labeledPoint: LabeledPoint): Boolean = !(labeledPoint.weight.isNaN || labeledPoint.weight.isInfinite)
 
   /**
+   * Validate a data set using one or more data point validators.
    *
-   * @param dataSet
-   * @param perSampleValidators
-   * @return
+   * @param dataSet The input data set
+   * @param perSampleValidators A list of (data validator, error message) pairs
+   * @return The list of validation error messages for the input data
    */
-  def validateFeatures(dataSet:RDD[LabeledPoint], perSampleValidators:Map[String, LabeledPoint=>Boolean]): Boolean = {
-    dataSet.mapPartitions(x => {
-      Seq(x.forall(item => {
-        perSampleValidators.map( validator => {
-          val valid = validator._2(item)
-          if (!valid) {
-            logger.error(s"Validation ${validator._1} failed on item: $item")
-          }
-          valid
-        }).forall(x => x)
-      })).iterator
-    }).fold(true)(_ && _)
-  }
+  private def validateData(
+      dataSet: RDD[LabeledPoint],
+      perSampleValidators: List[((LabeledPoint => Boolean), String)]): Seq[String] =
+    perSampleValidators
+      .map { case (validator, msg) =>
+        val validatorBroadcast = dataSet.sparkContext.broadcast(validator)
+        val result = dataSet.aggregate(true)(
```

Comment:
We have deliberated on and seen performance tradeoffs betwen aggregate vs treeAggregate and the depth of it but that was for the gradient aggregation which is called repeatedly during training. I think it should be fine here.